src/Dialogs/MainWindow.vala

```diff
@@ -198,17 +183,4 @@ public class Tuba.Dialogs.MainWindow: Adw.ApplicationWindow, Saveable {
 
 		last_view = view;
 	}
-
-	[GtkCallback]
-	void on_child_transition () {
-		if (leaflet.child_transition_running)
-			return;
-
-		Widget unused_child = null;
-		while ((unused_child = leaflet.get_adjacent_child (Adw.NavigationDirection.FORWARD)) != null) {
-			leaflet.remove (unused_child);
-			unused_child.dispose ();
-		}
-	}
-
```

Comment:
Memory leak

That callback would remove pages from the leaflet when you navigated backwards and then dispose them. For AdwNavigationView, the replacement would be a combination of Adw.NavigationView#popped(?) and Adw.NavigationPage#hidden. It'll be a bit tricky to implement as it first needs to be removed from the NavigationView (Adw.NavigationView#remove) before it gets disposed but only when Adw.NavigationPage#hidden is emitted (otherwise the transition won't play and the page will just disappear) (there might be better ways to do this than that however)

To check if something is memory leaking at a glance, look at the logs; going back *should* log a "Destroying <widget>" log for the view and all its children (e.g. in the current version of Tuba, opening the federated view and then going backwards should display a "destroying status" log for every post and some for the view

Reply:
I must be missing something fundamental here: wouldn't the navigation view pop (and so unreference) the view by itself when you go back? And then if nothing else references the view it gets destroyed. No?

Reply:
But indeed you're right and I see that the destructor doesn't get called. So is something else still referencing the view?

Reply:
> I must be missing something fundamental here: wouldn't the navigation view pop (and so unreference) the view by itself when you go back? And then if nothing else references the view it gets destroyed. No?

It *should*, but memory leaks occur for random reasons sometimes in Vala (https://github.com/search?q=repo%3AGeopJr%2FTuba+memory+leak&type=commits (especially with lambdas))

> So is something else still referencing the view?

That was added pre-Tuba likely for a similar reason (removing it from the leaflet should be disposing it too). Investigating it is probably not worth doing on this PR

Reply:
> memory leaks occur for random reasons sometimes in Vala (especially with lambdas))

Oh my, this is very very bad. And you're working around this by calling `g_object_run_dispose ()` when popping a page? You should definitely investigate where the leak/refcycle is coming from and find a proper way to prevent it, rather than doing it like this. But you've got a point about not doing this in this PR.

Reply:
OK, I pushed an attempt to do something like that. The navigation view emits `popped` _before_ actually unparenting the widget, so it doesn't work to `dispoe ()` it at that point. One would think that running `dispose ()` in the net iteration of the loop (`g_idle_add_once ()`) is the obvious solution, but it seems that `Idle.add_once ()` is even more broken/cursed in Vala: it causes things to segfault immediately, and the generated C code makes no sense.

So what I did that seems to work is in the widget itself, watch for it getting unparented, and run `dispose ()` at that point.

Reply:
Sounds good and looks much more elegant! Thanks!